ckan/lib/humanize.py

```diff
@@ -0,0 +1,141 @@
+# -*- coding: utf-8 -*-
+
+import logging
+from ckan.common import _
+
+log = logging.getLogger(__name__)
+
+
+def reset_humanizers():
+    for k in _instances:
+        _instances[k] = None
+
+    # First base will always be provided by current module. We want to
+    # keep it as fallback
+    for k, v in _bases.items():
+        _bases[k] = v[:1]
+
+
+def add_humanizer(humanizer_type, cls):
+    """Add custom class to humanizers hierarchy.
+
+    Class must be inherited from
+    :py:class:`~ckan.lib.humanize.BaseHumanizer`. Every time
+    ``h.humanize_entity_type`` is called, its third argument `purpose`
+    will be converted to method name by replacing all whitespaces with
+    underscore. If corresponding method exists in humanizer instance,
+    it will be used for obtaining required message.
+
+    Example::
+
+      class LinkHumanizer(BaseHumanizer):
+          def add_link(self, term, entity_type):
+              # term: custom group/org/package type, i.e, `camel-photos`
+              # entity_type: native entity name, i.e, `organization`,
+              # `group`, `dataset`.
+              return _("Create new %s" % term)
+
+      class ExtPlugin(p.SingletonPlugin):
+          p.implements(p.IConfigure)
+          def update_config(self, config_):
+              tk.add_humanizer('entity_type', LinkHumanizer)
+
+      # will be used in following case
+      h.humanize_entity_type('group', 'custom_group', 'add link')
+
+    :param humanizer_type: defines type of conversions performed by
+        humanizer. At the moment only ``entity_type`` supported.
+    :type humanizer_type: string
+
+    :param cls: Class that extends :py:class:`~ckan.lib.humanize.BaseHumanizer`
+    :type cls: class
+
+    """
+    _bases[humanizer_type].append(cls)
+
+
+def get_humanizer(type_, **kwargs):
+    if not _instances[type_]:
+        _instances[type_] = type(
+            type_, tuple(reversed(_bases[type_])), kwargs)()
+    return _instances[type_]
+
+
+class BaseHumanizer(object):
+    def _prepare_term(self, term):
+        return term.replace(u"_", u" ").capitalize()
+
+    def _default(self, term, **kwargs):
+        # Use default value from template
+        return
+
+    def __call__(self, purpose, term, *args, **kwargs):
+        handler = getattr(self, purpose.replace(u" ", u"_"), None)
+        term = self._prepare_term(term)
+        if handler:
+            return handler(term, *args, **kwargs)
+        log.debug(u'%s has no handlers for "%s"', self.__class__, purpose)
+        return self._default(term, **kwargs)
+
+
+class EntityTypeHumanizer(BaseHumanizer):
+    def add_link(self, term, entity_type):
+        return _(u"Add {}".format(term))
```

Comment:
These need to be written as `_(u"Add {}").format(term)` so that extract_messages can pick them up.

Reply:
This option works only with names\numbers and won't work with terms. 
Let's use `Add {type}` as example. In translation to japanes it will turn into ` {type}を追加`, which becomes `Camel photosを追加` after interpolation with `camel_photos` type.

Instead, let's extract just default messages from templates and leave those humanized labels. user will clearly see `Add Camel photos` in japanese locale and will manually add into .po file following lines:
```
msgid "Add Camel photos"
msgstr " 駱駝の写真を追加"
```

Reply:
It's ok that the translations with placeholders don't work for every language. The default humanize_entity_type is *best effort*. When you want an exact translation in your language you can name your type something that makes sense in your preferred language, or to support multiple languages override the helper or use an extension like ckanext-scheming and put all the message translation in your schema next to the rest of the descriptions of those schemas.

Reply:
Ok, makes sense. But i expect complains from people, who think that such functionality must be available out of the box :)